storage/database.go

```diff
@@ -203,25 +205,67 @@ func (d *db) Write(
 	return shard.Write(ctx, id, timestamp, value, unit, annotation)
 }
 
-func (d *db) ReadEncoded(
-	ctx context.Context,
-	id string,
-	start, end time.Time,
-) ([][]xio.SegmentReader, error) {
+func (d *db) readableShard(shardID uint32) (databaseShard, error) {
 	d.RLock()
 	if !d.bsm.IsBootstrapped() {
 		d.RUnlock()
 		return nil, errDatabaseNotBootstrapped
 	}
-	shardID := d.shardSet.Shard(id)
 	shard := d.shards[shardID]
 	d.RUnlock()
 	if shard == nil {
 		return nil, fmt.Errorf("not responsible for shard %d", shardID)
 	}
+	return shard, nil
+}
+
+func (d *db) ReadEncoded(
+	ctx context.Context,
+	id string,
+	start, end time.Time,
+) ([][]xio.SegmentReader, error) {
+	shardID := d.shardSet.Shard(id)
+	shard, err := d.readableShard(shardID)
+	if err != nil {
+		return nil, err
+	}
 	return shard.ReadEncoded(ctx, id, start, end)
 }
 
+func (d *db) FetchBlocks(
+	ctx context.Context,
+	id string,
+	starts []time.Time,
+) []FetchBlockResult {
+	shardID := d.shardSet.Shard(id)
+	shard, err := d.readableShard(shardID)
+	if err != nil {
+		// If we don't own the shard associated with the id, return an error
+		// so the client knows it needs to retry these.
```

Comment:
Sure did it this way because i wasn't sure if you were performing top-level checks. If you are, agreed we should just return a top-level error.